services/horizon/internal/db2/history/transaction.go

```diff
@@ -11,21 +12,59 @@ import (
 	"github.com/stellar/go/xdr"
 )
 
+// TransactionsWithoutPrefixExist returns true if there is at least one
+// transaction in history_transactions table without transaction_hash_prefix
+// field set.
+func (q *Q) TransactionsWithoutPrefixExist(ctx context.Context) (bool, error) {
+	var id int64
+	err := q.GetRaw(ctx, &id, "SELECT id FROM history_transactions WHERE transaction_hash_prefix IS NULL LIMIT 1")
+	if err != nil && err != sql.ErrNoRows {
+		return false, errors.Wrap(err, "error calculating number of nulls in transaction_hash_prefix")
+	}
+
+	// No err means at least one tx with no prefix set exists
+	return err == nil, nil
+}
+
 // TransactionByHash is a query that loads a single row from the
 // `history_transactions` table based upon the provided hash.
 func (q *Q) TransactionByHash(ctx context.Context, dest interface{}, hash string) error {
-	byHash := selectTransaction.
-		Where("ht.transaction_hash = ?", hash)
-	byInnerHash := selectTransaction.
-		Where("ht.inner_transaction_hash = ?", hash)
-
-	byInnerHashString, args, err := byInnerHash.ToSql()
-	if err != nil {
-		return errors.Wrap(err, "could not get string for inner hash sql query")
+	// Check if all rows in `history_transactions` have prefix field set. If so
+	// let's use the index on `transaction_hash_prefix` because index on
+	// `transaction_hash` could be removed by Horizon admins for performance
+	// reasons (faster ingestion).
+	txWithoutPrefixExist, outerErr := q.TransactionsWithoutPrefixExist(ctx)
```

Comment:
Would this code be easier to read if it had fewer double-negatives?
```suggestion
	txPrefixIndexReady, outerErr := q.TransactionPrefixIndexReady(ctx)
```
(ready/available/ok/whatever)

Then below you could do:
```go
if txPrefixIndexReady {
  // All transactions have prefix field set -> use new index
  // ...
} else {
  // no index
}